WIP - investigate reducing allocations in conversion

smarterclayton opened the pull request:
Conversion of a pod is ~40 allocations in each direction.  If we don't need to preserve the existing struct, we can assign equivalent types and avoid allocations.

Mostly just trying out things - going to generate direct defaulters next to verify that we can cut out a bunch more allocations on both sides.  

@wojtek-t as per our discussions

smarterclayton committed:
Avoid allocations and a reflect.Call in conversion

reflect.Call is fairly expensive, performing 8 allocations and having to
set up a call stack. Using a fairly straightforward to generate switch
statement, we can bypass that early in conversion (as long as the
function takes responsibility for invocation). We may also be able to
avoid an allocation for the conversion scope, but not positive yet.

```
benchmark                    old ns/op     new ns/op     delta
BenchmarkPodConversion-8     14713         12173         -17.26%

benchmark                    old allocs     new allocs     delta
BenchmarkPodConversion-8     80             72             -10.00%

benchmark                    old bytes     new bytes     delta
BenchmarkPodConversion-8     9133          8712          -4.61%
```

smarterclayton committed:
Conversions

smarterclayton committed:
Experiment - ref, not copy conversions

smarterclayton committed:
Codecgen

smarterclayton committed:
Convert remaining int types to fixed

smarterclayton committed:
Fix the rest of the code

smarterclayton committed:
Direct copy equal maps

smarterclayton committed:
Generated post map

smarterclayton committed:
Experiment - use type aliases for equivalent structs

Instead of duplicating tons of structs, use Go type aliases to identify
equivalent structs. In generators, try converting the objects directly.

Blocked by a bug in go2idl - alias information isn't preserved by the
time it hits the generator, so we can't determine equivalence correctly.

bgrant0607 commented:
cc @kubernetes/sig-api-machinery

bgrant0607 commented:
You know, you can assign PRs when you create them. :-)

smarterclayton commented:
I use "hub" from the CLI and it doesn't have that option, or i would :)

smarterclayton commented:
@wojtek-t @lavalamp when trying to use type aliases in the last commit (v1.ContainerStateRunning is a Go alias to api.ContainerStateRunning), by the time I get into the conversion generator types.Type for the left hand is not of Kind=Alias (it thinks left and right are two distinct structs).  Do you have any idea where in go2idl we'd be resolving the Alias reference aggressively?

```
doStruct(
  &types.Type{Name:types.Name{Package:"k8s.io/kubernetes/pkg/api/v1", Name:"ContainerStateRunning", Path:""}, Kind:"Struct", CommentLines:"", SecondClosestCommentLines:"// ContainerStateRunning is a running state of a container.\ntype ContainerStateRunning struct {\n\t// Time at which the container was last (re-)started\n\tStartedAt unversioned.Time `json:\"startedAt,omitempty\" protobuf:\"bytes,1,opt,name=startedAt\"`\n}\n", Members:[]types.Member{types.Member{Name:"StartedAt", Embedded:false, CommentLines:"", Tags:"json:\"startedAt,omitempty\"", Type:(*types.Type)(0xc82b59a580)}}, Elem:(*types.Type)(nil), Key:(*types.Type)(nil), Underlying:(*types.Type)(nil), Methods:[]*types.Type{(*types.Type)(0xc82b3e8580), (*types.Type)(0xc82b3cfa20), (*types.Type)(0xc82b3e8580), (*types.Type)(0xc82b402370), (*types.Type)(0xc82b402580), (*types.Type)(0xc82b402630), (*types.Type)(0xc82b4026e0), (*types.Type)(0xc82b4558c0), (*types.Type)(0xc82b529970), (*types.Type)(0xc82b529a20), (*types.Type)(0xc82b529a20), (*types.Type)(0xc82b454b00)}, Signature:(*types.Signature)(nil)}, 
  &types.Type{Name:types.Name{Package:"k8s.io/kubernetes/pkg/api", Name:"ContainerStateRunning", Path:""}, Kind:"Struct", CommentLines:"", SecondClosestCommentLines:"", Members:[]types.Member{types.Member{Name:"StartedAt", Embedded:false, CommentLines:"", Tags:"json:\"startedAt,omitempty\"", Type:(*types.Type)(0xc82b59a580)}}, Elem:(*types.Type)(nil), Key:(*types.Type)(nil), Underlying:(*types.Type)(nil), Methods:[]*types.Type{(*types.Type)(0xc82b4558c0), (*types.Type)(0xc82b529970), (*types.Type)(0xc82b529a20), (*types.Type)(0xc82b529a20)}, Signature:(*types.Signature)(nil)}
)
```

smarterclayton commented:
Reusing ints, maps, and the existing reflect call (linked PR) improved conversion time by 33% just by removing ~14 allocations (out of 80 for convert in and convert out).

The previous optimization we discussed for defaulting (avoiding reflection and inlining the method calls, like we did for conversions) looks like the biggest low hanging fruit - I think we can potentially remove ~2-3 allocations per defaulter, which should be a massive speed up there (possibly 35% of all allocations).

wojtek-t commented on cmd/kube-proxy/app/options/options.go:
```diff
@@ -63,25 +63,25 @@ func NewProxyConfig() *ProxyServerConfig {
 	fs.Var(componentconfig.IPVar{Val: &s.BindAddress}, "bind-address", "The IP address for the proxy server to serve on (set to 0.0.0.0 for all interfaces)")
 	fs.StringVar(&s.Master, "master", s.Master, "The address of the Kubernetes API server (overrides any value in kubeconfig)")
-	fs.IntVar(&s.HealthzPort, "healthz-port", s.HealthzPort, "The port to bind the health check server. Use 0 to disable.")
+	fs.Int32Var(&s.HealthzPort, "healthz-port", s.HealthzPort, "The port to bind the health check server. Use 0 to disable.")
```
@smarterclayton - all the changes int<->int32 are pretty independent and we can do them very safely.
Can you please move them to a separate PR - so that we can merge it earlier, since this doesn't seem to be controversial.

smarterclayton replied:
Yeah, had planned on it.  Wanted to get a total number first.

On Wed, Apr 27, 2016 at 2:46 AM, Wojciech Tyczynski <
notifications@github.com> wrote:

> In cmd/kube-proxy/app/options/options.go
> <https://github.com/kubernetes/kubernetes/pull/24845#discussion_r61211204>
> :
>
> > @@ -63,25 +63,25 @@ func NewProxyConfig() *ProxyServerConfig {
> >  func (s *ProxyServerConfig) AddFlags(fs *pflag.FlagSet) {
> >  	fs.Var(componentconfig.IPVar{Val: &s.BindAddress}, "bind-address", "The IP address for the proxy server to serve on (set to 0.0.0.0 for all interfaces)")
> >  	fs.StringVar(&s.Master, "master", s.Master, "The address of the Kubernetes API server (overrides any value in kubeconfig)")
> > -	fs.IntVar(&s.HealthzPort, "healthz-port", s.HealthzPort, "The port to bind the health check server. Use 0 to disable.")
> > +	fs.Int32Var(&s.HealthzPort, "healthz-port", s.HealthzPort, "The port to bind the health check server. Use 0 to disable.")
>
> @smarterclayton <https://github.com/smarterclayton> - all the changes
> int<->int32 are pretty independent and we can do them very safely.
> Can you please move them to a separate PR - so that we can merge it
> earlier, since this doesn't seem to be controversial.
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly or view it on GitHub
> <https://github.com/kubernetes/kubernetes/pull/24845/files/6f558a4578868b1c76783f147b6b34817efa5597#r61211204>
>

smarterclayton committed:
Inline default public methods

smarterclayton commented:
One more commit - inlined the defaulters (which is a huge source of allocations).  Some design points I'll need to go over:

```
# Initial (master)
BenchmarkPodConversion-8	  500000	     16362 ns/op	    9133 B/op	      80 allocs/op

# Final result (avoiding copies, inlining defaults, and the other PR for reflect)
BenchmarkPodConversion-8	 1000000	      6551 ns/op	    4040 B/op	      34 allocs/op
```

So ~60% speed up in conversion total, vs some worthwhile tradeoffs.

smarterclayton commented:
Found another 4 for pods

```
BenchmarkPodConversion-8	 1000000	      5584 ns/op	    3544 B/op	      30 allocs/op
```

smarterclayton committed:
Fixup hand created config

smarterclayton commented:
It suddenly occurs to me that our conversion framework is really only useful when types are exact (if they aren't, you can hand-roll your own, which you have to test and support them).  Which makes the alias problem more desirable to solve

smarterclayton committed:
Optimize group version allocations

smarterclayton committed:
Assign equivalent fields, not copy

smarterclayton committed:
MinReplicas in hand conversion needs to be changed

smarterclayton committed:
Generation of direct assignment

smarterclayton commented:
Last set

```
BenchmarkPodConversion-8                  	  300000	      5266 ns/op	    3288 B/op	      26 allocs/op
```

Noticed that QuantityCopy is insanely expensive, will fix this:

```
   938024   10981826 (flat, cum) 65.39% of Total
         .          .    510:func Convert_v1_ResourceList_To_api_ResourceList(in *ResourceList, out *api.ResourceList, s conversion.Scope) error {
         .          .    511:	if *in == nil {
         .          .    512:		return nil
         .          .    513:	}
         .          .    514:
    513388     513388    515:	converted := make(api.ResourceList)
         .          .    516:	for key, val := range *in {
         .    2545086    517:		value := val.Copy()
         .          .    518:
         .          .    519:		// TODO(#18538): We round up resource values to milli scale to maintain API compatibility.
         .          .    520:		// In the future, we should instead reject values that need rounding.
         .          .    521:		const milliScale = 3
         .    7498716    522:		value.Amount.Round(value.Amount, milliScale, inf.RoundUp)
         .          .    523:
    424636     424636    524:		converted[api.ResourceName(key)] = *value
```

smarterclayton commented:
Convert_v1_ResourceList_To_api_ResourceList is a big chunk of the remaining allocations because we have to round all the old resource versions.  May be an opportunity to avoid rounding in some cases.

smarterclayton commented:
@kubernetes/sig-scalability I've got some PRs coming out of this that should be significant CPU and memory reductions on the masters - this combined with protobuf should dramatically reduce overall load and help us hit the next tier.  If anyone has review bandwidth I can use as many reviews as you have time.

timothysc commented:
just @ the sig on the PRs and we'll take.

davidopp commented:
For now @wojtek-t is the best first-line reviewer for these PRs, though I'm sure he'll delegate if he thinks it is warranted.

lavalamp commented:
This is great stuff, I don't know how you find the time to do it. I think @wojtek-t will be out for a couple days due to holiday in Poland. I can try to review things in the mean time.

smarterclayton commented:
Closing this - all the things here are in other PRs.  Total reduction looks like 50-75% of allocations and time spent in conversion, but still need to get protobuf integrated so we can see the combined impact (~60us -> 15us in should be possible on the encode side, ~60 -> 25us should be possible on the decode side).
